adapter/stackdriver/stackdriver.go

```diff
@@ -0,0 +1,312 @@
+				Type: "global",
+				Labels: map[string]string{
+					"project_id": s.projectID,
+				},
+			},
+			MetricKind: info.kind,
+			ValueType:  info.value,
+			// Since we're sending a `CreateTimeSeries` request we can only populate a single point, see
+			// the documentation on the `points` field: https://cloud.google.com/monitoring/api/ref_v3/rest/v3/TimeSeries
+			Points: []*monitoringpb.Point{{
+				Interval: &monitoringpb.TimeInterval{
+					StartTime: start,
+					EndTime:   end,
+				},
+				Value: toTypedVal(val.MetricValue, val.Definition.Value)},
+			},
+		})
+	}
+	s.m.Lock()
+	s.toSend = append(s.toSend, data...)
+	s.m.Unlock()
+	return nil
+}
+
+func (s *sd) pushData() {
+	l := s.env.Logger()
+
+	l.Infof("Pushing data to Stackdriver")
+	s.m.Lock()
+	if len(s.toSend) == 0 {
+		s.m.Unlock()
+		l.Infof("No data to send to Stackdriver")
+		return
+	}
+	// Take the ref to the data we're pushing and create a new one to be written in to. We assume it'll be similarly
+	// sized to the last one.
+	// TODO: evaluate just swapping between two arrays (old, new) rather than creating new ones. Could run into
+	// problems if latency is high (> cfg.PushInterval) due to sending a lot of data in one go. Otherwise, maybe pool things
+	toSend := s.toSend
+	s.toSend = make([]*monitoringpb.TimeSeries, 0, len(toSend))
+	s.m.Unlock()
+	l.Infof("%d timeseries before merging", len(toSend))
+	// Combine timeseries with adjacent/overlapping intervals to ensure no times collide in this batch for DELTA metric series.
+	toSend = coalesce(toSend, l)
+	l.Infof("%d timeseries after merging", len(toSend))
+
+	err := s.pushMetrics(context.Background(),
+		&monitoringpb.CreateTimeSeriesRequest{
+			Name:       monitoring.MetricProjectPath(s.projectID),
+			TimeSeries: toSend,
+		})
+
+	// TODO: this is executed in a daemon, so we can't get out info about errors other than logging.
+	// We need to build framework level support for these kinds of async tasks. Perhaps a generic batching adapter
+	// can handle some of this complexity?
+	l.Infof("Stackdriver returned: %v", err)
+}
+
+func (s *sd) Close() error {
+	return s.client.Close()
+}
+
+func toStringMap(in map[string]interface{}) map[string]string {
+	out := make(map[string]string, len(in))
+	for key, val := range in {
+		out[key] = fmt.Sprintf("%v", val)
+	}
+	return out
+}
+
+func toTypedVal(val interface{}, t adapter.LabelType) *monitoringpb.TypedValue {
+	switch labelMap[t] {
+	case labelpb.LabelDescriptor_BOOL:
+		return &monitoringpb.TypedValue{&monitoringpb.TypedValue_BoolValue{BoolValue: val.(bool)}}
+	case labelpb.LabelDescriptor_INT64:
+		if t, ok := val.(time.Time); ok {
+			val = t.Nanosecond() / int(time.Microsecond)
+		} else if d, ok := val.(time.Duration); ok {
+			val = d.Nanoseconds() / int64(time.Microsecond)
+		}
+		return &monitoringpb.TypedValue{&monitoringpb.TypedValue_Int64Value{Int64Value: val.(int64)}}
+	case labelpb.LabelDescriptor_STRING:
+		fallthrough
```

Comment:
if it is a string, could we save some time by avoiding the `fmt.Sprintf` ?  maybe premature at this point.